### PR TITLE
add statful modules guide

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

WIP

### CODE REVIEW

File: docs/guides/stateful_modules.rst [hunk 1]
```diff
@@ -0,0 +1,28 @@
+Using statful modules
```
Comment (marcvanzee) [inline docs/guides/stateful_modules.rst:1]: I wonder whether we should call it "Stateful Modules" or "Bound Modules". I think Stateful Modules are a bit confusing since Flax Modules are generally stateless (and we say this *a lot*).

Also, as a general comment, I wonder whether we need to add a verb before every guide, or whether it would be clearer to just use the noun phrase. So "Using Bound Modules" --> "Bound Modules", "Applying batch normalization" --> "Batch Normalization", etc. WDYT?

```suggestion
Using stateful modules
```

**Comment by cgarciae** [comment]
Currently `bind` fails for cases like this:

```python
import flax.linen as nn
import jax
import jax.numpy as jnp

module = nn.Sequential([
    nn.Sequential([
        nn.Dense(8),
        nn.relu,
        nn.Dense(4),
    ]),
    nn.relu,
    nn.Dense(4),
])
x = jnp.ones((1, 5))
variables = module.init(jax.random.PRNGKey(0), x)
bound_module = module.bind(variables)

print(bound_module.layers[0].layers[0].scope) # None  <== !!!
```
@levskaya will try to come up with a solution to bind submodules, this guide will have to wait for that to be resolved.

File: docs/guides/stateful_modules.rst [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+   * Show that you can access submodules defined in ``setup``.
+
+   * Show how to use ``bind`` + ``unbind`` to extract a submodule
+   and its variables.
+#. Example of using ``bind`` for debbugging / inspecting the computation:
+
+   * Show how to use ``bind`` to manually run a ``Sequential`` module layer
+   by layer.
+
+#. Edge cases of using ``bind``:
```
[the remark this answers is not in the document]
Comment (8bitmp3) [reply docs/guides/stateful_modules.rst]: TODO: @8bitmp3
